File: src/main/java/com/github/quiltservertools/wires/StaffChat.java
package com.github.quiltservertools.wires;

import me.lucko.fabric.api.permissions.v0.Permissions;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.LiteralText;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.Util;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class StaffChat {
    private List<UUID> players = new ArrayList<>();

    public void toggle(UUID uuid) {
        if (!players.removeIf(uuid1 -> uuid1.equals(uuid))) {
            players.add(uuid);
        }
    }

    public boolean isInStaffChat(UUID uuid) {
        return this.players.stream().anyMatch(uuid1 -> uuid1.equals(uuid));
    }

    public void sendMessage(ServerPlayerEntity sender, String message) {
        Text text = new LiteralText("[Staff Chat] <" + sender.getName().asString() +"> " + message).formatted(Formatting.YELLOW);
        Objects.requireNonNull(sender.getServer()).getPlayerManager().getPlayerList().stream().filter(serverPlayerEntity -> Permissions.check(serverPlayerEntity, "wires.staffchat", 3)).forEach(player -> {
            player.sendSystemMessage(text, Util.NIL_UUID);
        });
    }
}

File: src/main/kotlin/net/quiltservertools/wires/StaffChat.kt
package net.quiltservertools.wires

import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.LiteralText
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import net.minecraft.util.Util
import net.quiltservertools.wires.util.Permissions
import net.quiltservertools.wires.util.Permissions.hasPermission
import java.util.*
import kotlin.collections.ArrayList

object StaffChat {

    private val players: MutableList<UUID> = ArrayList()
    private val name = "staffchat"

    fun toggle(uuid: UUID) {
        if (!players.removeIf { it == uuid }) {
            players.add(uuid)
        }
    }

    fun isInStaffChat(uuid: UUID): Boolean {
        return players.stream().anyMatch { it == uuid }
    }

    fun sendMessage(sender: ServerPlayerEntity, message: String) {
        val text: Text =
            LiteralText("[Staff Chat]").formatted(Formatting.GOLD).append(LiteralText(" <${sender.name.asString()}> $message").formatted(Formatting.WHITE))
        sender.getServer()?.playerManager!!.playerList!!.stream()
            .filter {
                it.commandSource.hasPermission(name)
            }.forEach {
                it.sendSystemMessage(
                    text,
                    Util.NIL_UUID
                )
            }
    }
}